pkg/identity/github/issuer_test.go

```diff
@@ -0,0 +1,176 @@
+	for name, test := range tests {
+		t.Run(name, func(t *testing.T) {
+			got := issuer.Match(ctx, test.URL)
+			if got != test.ShouldMatch {
+				t.Error(name)
+			}
+		})
+	}
+}
+
+type testKeySet struct {
+}
+
+func (t *testKeySet) VerifySignature(ctx context.Context, jwt string) ([]byte, error) {
+	// Doesn't actually verify the token, just returns the payload
+	parts := strings.Split(jwt, ".")
+	if len(parts) != 3 {
+		return nil, fmt.Errorf("jwt: must have 3 parts")
+	}
+	return base64.RawURLEncoding.DecodeString(parts[1])
+}
+
+func TestIssuerAuthentication(t *testing.T) {
+	tests := map[string]struct {
+		Token           string
+		ExpectPrincipal workflowPrincipal
+		WantErr         bool
+	}{
+		`Valid token authenticates with correct claims`: {
+			/* Example claims from https://awsteele.com/blog/2021/09/15/aws-federation-comes-to-github-actions.html
+			{
+			  "actor": "aidansteele",
+			  "aud": "sigstore",
+			  "base_ref": "",
+			  "event_name": "push",
+			  "exp": 1631672856,
+			  "head_ref": "",
+			  "iat": 1631672556,
+			  "iss": "https://token.actions.githubusercontent.com",
+			  "job_workflow_ref": "aidansteele/aws-federation-github-actions/.github/workflows/test.yml@refs/heads/main",
+			  "jti": "8ea8373e-0f9d-489d-a480-ac37deexample",
+			  "nbf": 1631671956,
+			  "ref": "refs/heads/main",
+			  "ref_type": "branch",
+			  "repository": "aidansteele/aws-federation-github-actions",
+			  "repository_owner": "aidansteele",
+			  "run_attempt": "1",
+			  "run_id": "1235992580",
+			  "run_number": "5",
+			  "sha": "bf96275471e83ff04ce5c8eb515c04a75d43f854",
+			  "sub": "repo:aidansteele/aws-federation-github-actions:ref:refs/heads/main",
+			  "workflow": "CI"
+			}
+			*/
+			// Token generated with jwt.io and the claims above
+			Token: `eyJhbGciOiJSUzI1NiIsImtpZCI6Imptczl3WWdPcmxUWjZsSzZTZ2FfaE1JNDZHcXJXYUhicWZfdHJuZ1NxbDAiLCJ0eXAiOiJKV1QifQ.eyJhY3RvciI6ImFpZGFuc3RlZWxlIiwiYXVkIjoic2lnc3RvcmUiLCJiYXNlX3JlZiI6IiIsImV2ZW50X25hbWUiOiJwdXNoIiwiZXhwIjoxLjYzMTY3Mjg1NmUrMDksImhlYWRfcmVmIjoiIiwiaWF0IjoxLjYzMTY3MjU1NmUrMDksImlzcyI6Imh0dHBzOi8vdG9rZW4uYWN0aW9ucy5naXRodWJ1c2VyY29udGVudC5jb20iLCJqb2Jfd29ya2Zsb3dfcmVmIjoiYWlkYW5zdGVlbGUvYXdzLWZlZGVyYXRpb24tZ2l0aHViLWFjdGlvbnMvLmdpdGh1Yi93b3JrZmxvd3MvdGVzdC55bWxAcmVmcy9oZWFkcy9tYWluIiwianRpIjoiOGVhODM3M2UtMGY5ZC00ODlkLWE0ODAtYWMzN2RlZXhhbXBsZSIsIm5iZiI6MS42MzE2NzE5NTZlKzA5LCJyZWYiOiJyZWZzL2hlYWRzL21haW4iLCJyZWZfdHlwZSI6ImJyYW5jaCIsInJlcG9zaXRvcnkiOiJhaWRhbnN0ZWVsZS9hd3MtZmVkZXJhdGlvbi1naXRodWItYWN0aW9ucyIsInJlcG9zaXRvcnlfb3duZXIiOiJhaWRhbnN0ZWVsZSIsInJ1bl9hdHRlbXB0IjoiMSIsInJ1bl9pZCI6IjEyMzU5OTI1ODAiLCJydW5fbnVtYmVyIjoiNSIsInNoYSI6ImJmOTYyNzU0NzFlODNmZjA0Y2U1YzhlYjUxNWMwNGE3NWQ0M2Y4NTQiLCJzdWIiOiJyZXBvOmFpZGFuc3RlZWxlL2F3cy1mZWRlcmF0aW9uLWdpdGh1Yi1hY3Rpb25zOnJlZjpyZWZzL2hlYWRzL21haW4iLCJ3b3JrZmxvdyI6IkNJIn0.BCQD4kedZDzF_3IphHH4pr6cTYQzO9d0orFocvqZSSWH6hHYMw3PJ5YlAleEuXNxgVmMtHN6USrxbfa3bXXshhluSXiV4TVlPc2s8fFXNiLg9TwbXFkDDm_IfYZInHeS5By2AifbT0MPCebPnhYvVR1nUIAjZnU1lcgiN20rZFCV`,
```

Comment:
What do you think about generating the tokens via code, like https://github.com/sigstore/fulcio/blob/main/pkg/api/grpc_server_test.go#L484, instead of hardcoding them? The benefits are that it's easier to see what the actual values of the token are, and we won't have someone else's tokens and pii in the codebase.

Reply:
Ah very nice. Yes that is much more readable. I'll refactor the test to make the claims more clear and generate the tokens on the fly. Friday afternoon brains were a little fried 😆